### PR TITLE
new python data science module gppkg

### PR DESCRIPTION
**Pull request by lisakowen**
the plpython data science module .gppkg is not yet complete.  wanted to get some early feedback on theis doc.  i am looking for a STRUCTURAL REVIEW of this content at this time, not an editorial review.  i will request that when there is a production .gppkg version to try out.

i restructured some of the information to better align the content:
- moved "verifying the python environment" info earlier in the doc.
- addressing both the .gppkg and custom modules in the "Installing Python Modules" section.  subsections for the .gppkg and custom module install
- the page included a few examples of building/installing custom modules.  one of these modules was numpy, which is in the new .gppkg. removed this example.
- created new topic for troubleshooting python module import failures.  could use more info here.
- removed some of the references
- will need to conditionalize some of these references

questions:
- how to identify the .gppkg in words?  this doc is referring to it as the "Greenplum Database Python Data Science Module Package", which is a mouthful.
- is the general restructure reasonable?
- what do you think of the table listing all of the modules and (TBD) descriptions, and its placement?

### CODE REVIEW

**Comment by mkiyama** [comment]
After a quick look...

For the package name, I would just call it the Greenplum Database PL/Python (Language) Package describe in the text that it included data science modules.

I would focus the install topic to just installing the package because the package includes the data science pl/python modules. The modules are included as a separate install.

I would put the list of packages closer to the end of the section and a link at the top in the first set of bullets.

Are the versions needed for the modules? 
Are links to module docs needed for list of functions?

**Comment by lisakowen** [comment]
thanks for reviewing, @mkiyama .  not sure if i understand what you are suggesting w.r.t. the package name and install topic:
- are you thinking that the "installing python modules" section include only installing/uninstalling the data science module package and have a separate section for the build/install custom modules?  
- i am all for a simpler name.  but calling it the "Greenplum Database PL/Python (Language) Package" is kind of misleading to me, because we have other other languages (R, java) that you do currently install with a gppkg.  where as with PL/Python, the language is installed by default and you are installing only the extra data science modules when you install the gppkg.

**Comment by mkiyama** [comment]
oh, geez. I forgot pl/python is already installed.
so the package just installs all the DS python modules.

OK. revised comment.
I still think the list of DS modules should go at the end.
Yes, I think there should be two sections, 
--installing/uninstalling the DS modules pkg.
--installing/uninstalling a python module

I don't how common it is or if it will make building the python module section
more confusing, but you might want to add some summary information about
requirements for a more complex python install. something like

"To build some python modules requires installing additional software
such as a third party library or a compiler. See the python module documentation
for information about the requirements for building the python module."

as for the package name, you might be able to get away with just saying Greenplum Database the first time and saying Python Data Science Modules Package the rest of the time.

**Comment by litzell** [comment]
I don't see gppkg addressed in the "Installing Python Modules" section, and I'm confused by the info about modules, gppkg, and using gpscp and gpssh. And then the first example is about setup_tools with no explanation about why you want setup_tools on every GP host. I think the way to deal with that is to define each of the actual alternatives for packing/installing downloadable and custom modules in one place, before introducing the GP data science modules. 

The info about developing on a single host and then building on all hosts is a little strange to me because it looks like we're suggesting using a segment host directly for development or admin, which is the first I've seen anything like that. I assume you need to develop in a GPDB environment so it would be on a master host. Can we find out what kind of development env is typical?

**Comment by lisakowen** [comment]
thanks for your comments @litzell .  i agree that lumping the data science module package and building/installing your own module into the same intro topic is confusing.  i will restructure to better delineate the cases.  with respect to the building/installing instructions for custom modules, i didn't modify that content at all, assumed it was correct.  i will ask for an engineering review of that section when the PR is ready for a content review.

File: gpdb-doc/dita/ref_guide/extensions/pl_python.xml [hunk 1]
```diff
@@ -285,299 +300,292 @@ $$;</codeblock>
...
-              <codeph>remote-hosts</codeph> lists all the remote Greenplum Database segment
-            hosts:<codeblock>gpssh -f remote_hosts
->>> tar -xzvf distribute-0.6.21.tar.gz
->>> cd setuptools-18.4
->>> python setup.py build &amp;&amp; python setup.py install
->>> python -c "import setuptools"
->>> exit</codeblock></li>
+        <p>Before you install the Python Data Science Module package, make sure that your Greenplum Database is running, you have sourced <codeph>greenplum_path.sh</codeph>, and that the <codeph>$MASTER_DATA_DIRECTORY</codeph> and <codeph>$GPHOME</codeph> environment variables are set.</p>
+        <ol>
+          <li>Download the Python Data Science module package from <xref href="https://network.pivotal.io/products/pivotal-gpdb" format="html" scope="external">Pivotal Network</xref>. Make note of the directory to which the file was downloaded.<p>The file name format of the package is <codeph>gpdbpython-&lt;version&gt;-relhel&lt;N&gt;-x86_64.gppkg</codeph>.</p></li>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_python.xml]: Make steps 1/2 generic or conditional - don't assume anything has been downloaded.

File: gpdb-doc/dita/ref_guide/extensions/pl_python.xml [hunk 2]
```diff
@@ -285,299 +300,292 @@ $$;</codeblock>
...
+          <li>Copy the downloaded file to the Greenplum Database master host.</li>
+          <li>Use the <codeph>gppkg</codeph> command to install the package. For example:<codeblock>$ gppkg -i gpdbpython-&lt;version&gt;-relhel&lt;N&gt;-x86_64.gppkg</codeblock><p><codeph>gppkg</codeph> installs the Python Data Science modules on all nodes in your Greenplum Database cluster. The command also updates the <codeph>PYTHONPATH</codeph>, <codeph>PATH</codeph>, and <codeph>LD_LIBRARY_PATH</codeph> environment variables in your <codeph>greenplum_path.sh</codeph> file.</p></li>
+          <li>Restart Greenplum Database. You must re-<codeph>source</codeph> <codeph>greenplum_path.sh</codeph> before restarting your Greenplum cluster:<codeblock>$ source /usr/local/greenplum-db/greenplum_path.sh
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_python.xml]: Doesn't seem like "source" should be in code format here.

**Comment by lisakowen** [comment]
ready for content review @dyozie, @mkiyama, @litzell if you are interested.

File: gpdb-doc/dita/ref_guide/extensions/pl_python.xml [hunk 3]
```diff
@@ -285,296 +300,288 @@ $$;</codeblock>
...
+        module, ensure that the appropriate software build dependencies
+        are installed and properly configured on the build system and that the build
+        generates the correct executable. For example, on a Linux system the build
+        should create a 64-bit executable.</p>
+      <note>Some Python modules require installing additional software, such as a third party library or a compiler. Refer to the Python module's documentation for information about the requirements for building and using the module.</note>
+      <p>When you install a custom Python module in Greenplum Database, you must add the module to
+        all segment and mirror hosts in the cluster. When you expand Greenplum Database, you
+        must add the Python modules to the new segment hosts. You can use the Greenplum Database
+        utilities <codeph>gpssh</codeph> and <codeph>gpscp</codeph> to run commands on,
+        and copy files between, Greenplum Database hosts.</p>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_python.xml]: I would say
"to run commands on, and copy files to, remote Greenplum Database hosts."

File: gpdb-doc/dita/ref_guide/extensions/pl_python.xml [hunk 4]
```diff
@@ -285,296 +300,288 @@ $$;</codeblock>
...
+                <entry colname="col1">Module Name</entry>
+                <entry colname="col2">Description/Used For</entry>
+              </row>
+            </thead>
+            <tbody>
+              <row>
+                <entry colname="col1">
+                  Beautiful Soup
+                </entry>
+                <entry colname="col2">navigating HTML and XML</entry>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_python.xml]: Almost hate to ask this, but can we initial-cap the Description blurbs in this table?

File: gpdb-doc/dita/ref_guide/extensions/pl_python.xml [hunk 5]
```diff
@@ -285,296 +300,288 @@ $$;</codeblock>
...
+      <body>
+        <p>Possible causes of a Python module import failure could include:
+        <ul id="ul_mod_import_fail">
+          <li>A problem accessing required libraries. For the NumPy example, Greenplum Database
+            might have a problem accessing the Python libraries on a
+            segment host.<p><i>Try</i>: Test importing the module directly on the
+            segment host. The following <codeph>gpssh</codeph> command tests importing
+              the NumPy module on the segment host named
+              <codeph>mdw1</codeph>.<codeblock>$ gpssh -h mdw1 -e 'python -c "import numpy"'</codeblock></p></li>
+          <li>Environment variables may not be configured in the Greenplum database environment. The Python <codeph>import</codeph> statement may not return an error in this case. <p><i>Try</i>: Ensure that any environment variables that are required by a module are properly set in <codeph>gpadmin</codeph>'s <codeph>.bashrc</codeph> file on every host in your Greenplum Database cluster.</p></li>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_python.xml]: Capitalise Greenplum **D**atabase.

**Pull request closed without merging**